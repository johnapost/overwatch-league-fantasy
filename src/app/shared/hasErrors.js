// @flow

export default (fieldsError: Object) =>
  Object.keys(fieldsError).some(field => fieldsError[field]);
