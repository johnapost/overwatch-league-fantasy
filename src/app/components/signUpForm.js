// @flow

import React, { Component } from 'react';
import { Card, Form, Input, Button, message } from 'antd';
import { compose } from 'redux';
import { isLoaded } from 'react-redux-firebase';
import { withRouter } from 'next/router';
import withFirestore from '../shared/withFirestore';
import hasErrors from '../shared/hasErrors';

type Props = {
  firestore: Object,
  form: Object,
  router: Object,
}

type State = {
  submittingForm: boolean
}

class SignUpForm extends Component<Props, State> {
  state = {
    submittingForm: false,
  }

  componentDidMount() {
    if (this.emailEl) {
      this.emailEl.focus();
    }
  }

  emailEl: ?HTMLElement = null

  handleSubmit = (e) => {
    e.preventDefault();
    const { firestore, form: { validateFieldsAndScroll }, router } = this.props;

    if (!isLoaded(firestore)) {
      return;
    }

    this.setState({ submittingForm: true });
    const closeMessage = message.loading('Signing up..', 0);

    const finishSubmitting = () => {
      this.setState({ submittingForm: false });
      closeMessage();
    };

    validateFieldsAndScroll((err, { email, password }) => {
      if (err) {
        finishSubmitting();
        return;
      }

      firestore
        .auth()
        .createUserWithEmailAndPassword(email, password)
        .then(() => {
          finishSubmitting();
          message.success('Successfully signed up!');
          router.push({
            pathname: '/login',
            query: { email },
          });
        })
        .catch(({ code, message: errorMessage }) => {
          finishSubmitting();
          message.error(errorMessage);
          // eslint-disable-next-line no-console
          console.error(code, message);
        });
    });
  }

  render() {
    const { form: { getFieldDecorator, getFieldsError } } = this.props;
    const { submittingForm } = this.state;
    const emailValidations = [
      { required: true, message: 'Email required!' },
      { type: 'email', message: 'Invalid email!' },
    ];
    const passwordValidations = [
      { required: true, message: 'Password required!' },
      { min: 6, message: 'Too short!' },
    ];

    return (
      <Card title="Sign up for an account">
        <Form onSubmit={this.handleSubmit}>
          <Form.Item>
            {
              getFieldDecorator(
                'email',
                { rules: emailValidations },
              )(<Input
                placeholder="Email"
                ref={(el) => { this.emailEl = el; }}
              />)
            }
          </Form.Item>
          <Form.Item>
            {
              getFieldDecorator(
                'password',
                { rules: passwordValidations },
              )(<Input placeholder="Password" type="password" />)
            }
          </Form.Item>
          <div style={{ marginTop: '10px' }}>
            <Button
              type="primary"
              htmlType="submit"
              disabled={submittingForm || hasErrors(getFieldsError())}
            >
              Sign Up
            </Button>
          </div>
        </Form>
      </Card>
    );
  }
}

export default compose(
  withRouter,
  withFirestore(),
  Form.create(),
)(SignUpForm);
