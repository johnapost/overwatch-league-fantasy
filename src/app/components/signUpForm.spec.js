import React from "react";
import { shallow } from "enzyme";
import { SignUpFormComponent } from "./signUpForm";
import hasErrors from "../shared/hasErrors";

jest.mock("../shared/hasErrors");

describe("SignUpForm", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    hasErrors.mockImplementation(() => false);
  });

  describe("component", () => {
    const props = {
      disabled: false,
      form: {
        getFieldDecorator: jest.fn(() => jest.fn(component => component)),
        getFieldsError: jest.fn()
      }
    };

    it("should render fine", () => {
      const wrapper = shallow(<SignUpFormComponent {...props} />);
      expect(wrapper.html()).toMatchSnapshot();
    });
  });
});
