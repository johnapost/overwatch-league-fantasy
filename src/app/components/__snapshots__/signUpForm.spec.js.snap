// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpForm component should render fine 1`] = `"<form class=\\"ant-form ant-form-horizontal\\"><div class=\\"ant-row ant-form-item\\"><div class=\\"ant-form-item-control-wrapper\\"><div class=\\"ant-form-item-control\\"><span class=\\"ant-form-item-children\\"><input type=\\"text\\" placeholder=\\"Email\\" value=\\"\\" class=\\"ant-input\\"/></span></div></div></div><div class=\\"ant-row ant-form-item\\"><div class=\\"ant-form-item-control-wrapper\\"><div class=\\"ant-form-item-control\\"><span class=\\"ant-form-item-children\\"><input type=\\"password\\" placeholder=\\"Password\\" value=\\"\\" class=\\"ant-input\\"/></span></div></div></div><div class=\\"ant-row ant-form-item\\"><div class=\\"ant-form-item-control-wrapper\\"><div class=\\"ant-form-item-control\\"><span class=\\"ant-form-item-children\\"><input type=\\"text\\" placeholder=\\"Display name\\" value=\\"\\" class=\\"ant-input\\"/></span></div></div></div><div style=\\"margin-top:10px\\"><button type=\\"submit\\" class=\\"ant-btn ant-btn-primary\\"><span>Sign Up</span></button></div></form>"`;
